src/NetMQ/Core/Patterns/XPub.cs

```diff
@@ -158,34 +166,41 @@ protected override void XReadActivated(Pipe pipe)
             while (pipe.Read(ref sub))
             {
                 // Apply the subscription to the trie.
-                byte[] data = sub.Data;
                 int size = sub.Size;
-                if (size > 0 && (data[0] == 0 || data[0] == 1))
+                if (size > 0 && (sub[0] == 0 || sub[0] == 1))
                 {
                     if (m_manual)
                     {
                         m_lastPipe = pipe;
 
-                        m_pending.Enqueue(sub.CloneData());
+                        m_pending.Enqueue(sub);
                     }
                     else
                     {
-                        var unique = data[0] == 0 
-                            ? m_subscriptions.Remove(data, 1, size - 1, pipe) 
-                            : m_subscriptions.Add(data, 1, size - 1, pipe);
+                        var unique = sub[0] == 0 
+                            ? m_subscriptions.Remove(sub.Data, sub.Offset + 1, size - 1, pipe) 
+                            : m_subscriptions.Add(sub.Data, sub.Offset + 1, size - 1, pipe);
 
                         // If the subscription is not a duplicate, store it so that it can be
                         // passed to used on next recv call.
                         if (m_options.SocketType == ZmqSocketType.Xpub && (unique || m_verbose))
-                            m_pending.Enqueue(sub.CloneData());
+                            m_pending.Enqueue(sub);
                     }
                 }
+                else if (m_broadcastEnabled && size > 0 && sub.HasMore && sub[0] == 2)
+                {
+                    m_lastPipe = pipe;
+                    m_lastPipeIsBroadcast = true;
+                    sub.Offset = sub.Offset + 1;
+                    sub.Size = sub.Size - 1;
+                    m_pending.Enqueue(sub);
+                    //XSend(ref sub);
+                }
                 else // process message unrelated to sub/unsub
                 {
-                    m_pending.Enqueue(sub.CloneData());
+                    m_pending.Enqueue(sub);
                 }
 
-                sub.Close();
```

Comment:
why you are not closing the Msg? there is cases the messages should be closed (it is not always enqueued)